### DISCUSSION TITLE
eth_newPendingTransactionFilter not working

### DISCUSSION THREAD

**Discussion by kolesnevg** [zkSync-Community-Hub/zksync-developers#180] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

Hi, I'm trying to find the fastest way to get pending transactions, I tried use eth_newPendingTransactionFilter to create filter and then use it via eth_getFilterChanges, but often I got error like 
`{"jsonrpc":"2.0","error":{"code":-32602,"message":"Filter not found"},"id":2}`
sometimes it works well, but most time it's not working at all(

Could you help me wit hthis task, may be I do something wrong.

Thank you

### What is the specific issue or error you're encountering?

above

### Can you share the error messages or logs you're receiving, if any?

`wscat --connect wss://mainnet.era.zksync.io/ws
Connected (press CTRL+C to quit)
> {"jsonrpc":"2.0","method":"eth_newPendingTransactionFilter","params": [],"id":1}
< {"jsonrpc":"2.0","result":"0x3c5b4b931828283e72ca248b3fb7324c11723843bdc549d6cfb99869844cbf59","id":1}
> {"jsonrpc": "2.0","method": "eth_getFilterChanges", "params": ["0x3c5b4b931828283e72ca248b3fb7324c11723843bdc549d6cfb99869844cbf59"], "id": 2}
< {"jsonrpc":"2.0","error":{"code":-32602,"message":"Filter not found"},"id":2}
`

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

no

### Can you provide the relevant portions of your contract code where the issue is occurring?

no contract

### Have you tried to isolate the problem, and if so, what were the results?

no

### What steps have you already taken to try to resolve the issue?

I tried use different servers, I thought that the problem is on Cloudflare side, but without result

**Comment by vasyl-ivanchuk** [comment, marked as answer]
Hey @kolesnevg 👋 ,

To prevent potential abuse by bad actors repeatedly creating filters, there is a limit on the number of filters that can be created. Because new filters override the old ones, it's possible that your filter may be overridden between your requests. I would recommend using `eth_subscribe` to receive updates on new pending transactions instead:

```
wscat -c wss://mainnet.era.zksync.io/ws
```
```
> {"jsonrpc":"2.0","id": 2, "method": "eth_subscribe", "params": ["newPendingTransactions"]}
```
please, let us know if it helps.

**Comment by patelkaren** [comment]
It seems like you are encountering intermittent issues with the eth_getFilterChanges RPC call returning "Filter not found" error. This error typically occurs when the filter ID provided to eth_getFilterChanges does not exist or has expired.

Here are some troubleshooting steps you can try to resolve this issue:

Ensure that you are using the correct filter ID returned by eth_newPendingTransactionFilter. Sometimes, filters can expire, so you may need to create a new filter periodically.

Check if there are any network disruptions or server-side issues that could be causing intermittent failures. You mentioned trying different servers, but if the problem persists, it might be beyond your control.

Consider using alternative methods for tracking pending transactions, such as subscribing to new block headers and extracting pending transactions from them. This approach can be more reliable than using filters.

Double-check your WebSocket connection to ensure that it remains stable and does not get disconnected unexpectedly.

Monitor the API endpoint's response times and error rates to identify any patterns or anomalies that could indicate underlying issues with the service provider.

By following these steps and potentially exploring alternative approaches, you may be able to resolve the issue with fetching pending transactions more reliably.

**Comment by clevelandmarcus** [comment]
"jsonrpc": "2.0",
  "error": {
    "code": -32602,
    "message": "Filter not found"
  },
  "id": 2

This error typically occurs when the filter ID provided to eth_getFilterChanges does not exist. It's possible that the filter expired or was invalidated due to inactivity. To ensure consistent results, consider implementing a retry mechanism in your application to handle cases where the filter is not found. Additionally, you may explore alternative approaches for tracking pending transactions, such as subscribing to new block headers and inspecting the transactions in the pending block.

**Comment by golfdutoresste** [comment]
To efficiently retrieve pending transactions, consider using Web3.js library's `eth_subscribe` method with the `newPendingTransactions` subscription type. This provides real-time updates on pending transactions without the need for filters, offering a more reliable and efficient solution.

**Comment by caojeremyv** [comment]
The user is trying to find the fastest way to get pending transactions on zkSync's Mainnet. They have tried using eth_newPendingTransactionFilter to create a filter and then eth_getFilterChanges to retrieve changes, but often receive the error "Filter not found". The user has not made any recent changes to the contract, and there are no external libraries or contracts being used. They have not provided any relevant code or attempted to isolate the problem. The user has tried using different servers to no avail.
